packages/core/plugins/core-plugins-contracts-browser/src/contracts.ts

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import type { PluginName } from '@kbn/core-base-common';
+
+/**
+ * Setup contract of Core's `plugins` service.
+ *
+ * @public
+ */
+export interface PluginsServiceSetup {
+  /**
+   * Returns a promise that will resolve with the requested plugin setup contracts once all plugins have been setup.
+   *
+   * @remark The execution order is not guaranteed to be consistent. Only guarantee is that the returned promise will be
+   *         resolved once all plugins are setup, and before Core's `start` is initiated.
+   *
+   * @remark If called when plugins are already setup, the returned promise will resolve instantly.
+   *
+   * @remark This API **should only be used** when trying to address cyclic dependency issues that can't be solved otherwise.
+   *
+   * @remark Requested dependencies must be defined as `runtimePluginDependencies` in the plugin's manifest,
+   *         otherwise the API will throw when called.
+   *
+   * @example
+   * ```ts
+   * setup(core) {
+   *   core.plugins.onSetup<{pluginA: SetupContractA, pluginB: SetupContractA}>('pluginA', 'pluginB')
+   *       .then(({ pluginA, pluginB }) => {
+   *         if(pluginA.found && pluginB.found) {
+   *           // do something with pluginA.contract and pluginB.contract
+   *         }
+   *       });
+   * }
+   * ```
+   */
+  onSetup: PluginContractResolver;
```

Comment:
FMI, will we throw if you specify a non-existant plugin in your `runtimePluginDependencies` list?

Reply:
```
@remark Requested dependencies must be defined as `runtimePluginDependencies` in the plugin's manifest,
        otherwise the API will throw when called.
```

Reply:
Oh you mean will we throw if there's an unknown plugin ID in the `runtimePluginDependencies` field of the manifest? No, I don't think this is necessary (and may even be dangerous with the way we toggle plugins in serverless?)

Reply:
yeah, that's a valid point! I was just thinking of catching typos, but the flexibility provided here does mean we cannot know what is/isn't a plugin ID.